arch/xtensa/core/mpu.c

```diff
@@ -0,0 +1,900 @@
+};
+
+/* No longer needed so undefine it. */
+#undef XCHAL_MPU_BGMAP
+
+/**
+ * Static definition of all code and data memory regions of the
+ * current Zephyr image. This information must be available and
+ * need to be processed upon MPU initialization.
+ */
+static const struct xtensa_mpu_range mpu_zephyr_ranges[] = {
+	/* Region for vector handlers. */
+	{
+		.start = (uintptr_t)XCHAL_VECBASE_RESET_VADDR,
+		/*
+		 * There is nothing from the Xtensa overlay about how big
+		 * the vector handler region is. So we make an assumption
+		 * that vecbase and .text are contiguous.
+		 *
+		 * SoC can override as needed if this is not the case,
+		 * especially if the SoC reset/startup code relocates
+		 * vecbase.
+		 */
+		.end   = (uintptr_t)__text_region_start,
+		.access_rights = XTENSA_MPU_ACCESS_P_RX_U_RX,
+		.memory_flags = XTENSA_MPU_MEMFLAGS_DEFAULT,
+		.name = "vecbase",
+	},
+	/*
+	 * Mark the zephyr execution regions (data, bss, noinit, etc.)
+	 * cacheable, read / write and non-executable
+	 */
+	{
+		/* This includes .data, .bss and various kobject sections. */
+		.start = (uintptr_t)_image_ram_start,
+		.end   = (uintptr_t)_image_ram_end,
+		.access_rights = XTENSA_MPU_ACCESS_P_RW_U_NA,
+		.memory_flags = XTENSA_MPU_MEMFLAGS_DEFAULT,
+		.name = "data",
+	},
+#if K_HEAP_MEM_POOL_SIZE > 0
+	/* System heap memory */
+	{
+		.start = (uintptr_t)_heap_start,
+		.end   = (uintptr_t)_heap_end,
+		.access_rights = XTENSA_MPU_ACCESS_P_RW_U_NA,
+		.memory_flags = XTENSA_MPU_MEMFLAGS_DEFAULT,
+		.name = "heap",
+	},
+#endif
+	/* Mark text segment cacheable, read only and executable */
+	{
+		.start = (uintptr_t)__text_region_start,
+		.end   = (uintptr_t)__text_region_end,
+		.access_rights = XTENSA_MPU_ACCESS_P_RX_U_RX,
+		.memory_flags = XTENSA_MPU_MEMFLAGS_DEFAULT,
+		.name = "text",
+	},
+	/* Mark rodata segment cacheable, read only and non-executable */
+	{
+		.start = (uintptr_t)__rodata_region_start,
+		.end   = (uintptr_t)__rodata_region_end,
+		.access_rights = XTENSA_MPU_ACCESS_P_RO_U_RO,
+		.memory_flags = XTENSA_MPU_MEMFLAGS_DEFAULT,
+		.name = "rodata",
+	},
+};
+
+/**
+ * Copy one MPU entry in an array to another entry.
+ *
+ * This copies all bits from MPU entry indexed @a from_idx to @a to_idx,
+ * and also change the segment field of the destination entry to reflect
+ * the new index.
+ *
+ * @param entries Array of MPU entries.
+ * @param from_idx Index of MPU entry to be copied from.
+ * @param to_idx Index of MPU entry to be copied to.
+ */
+static inline
+void mpu_entry_copy(struct xtensa_mpu_entry *entries, uint8_t from_idx, uint8_t to_idx)
+{
+	entries[to_idx] = entries[from_idx];
+	entries[to_idx].at.p.segment = to_idx;
+}
+
+/**
+ * Shift a portion of MPU entries in the array down one slot.
+ *
+ * This makes a hole in the MPU entry array so that a new entry can be inserted
+ * into the one indexed at @a start_idx. The entries [@a start_idx, @a end_idx - 1]
+ * will be moved to [@a start_idx + 1, @a end_idx].
```

Comment:
Is that correct ? Isn't it going down copying `end_idx` to `end_idx - 1`  ... resulting in `[start_idx, end_idx - 1]` ?

Reply:
The concept is kind of like stack where lower index is on top. So moving down means copying `end_idx -1` to `end_idx` (overwriting `end_idx`). `start_idx` will be a duplicate of `start_idx + 1`.